<!DOCTYPE html> <html lang="en">
<head>
	<title>MoDMaps3D - Main Menu</title>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">
	<script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script> 
	<script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>
	<!-- <link rel="stylesheet" href="js/bootstrap.min.css">
	<script src="js/jquery.min.js"></script>
	<script src="js/bootstrap.min.js"></script> -->


	<!-- Global site tag (gtag.js) - Google Analytics -->
	<script async src="https://www.googletagmanager.com/gtag/js?id=G-R0H3YXMEBS"></script>
	<script>
	window.dataLayer = window.dataLayer || [];
	function gtag(){dataLayer.push(arguments);}
	gtag('js', new Date());
	gtag('config', 'G-R0H3YXMEBS');
	</script>


	<script type="text/javascript" src="js/numeric-1.2.6.min.js"></script>
	<script type="text/javascript" src="js/jszip.min.js"></script>
	<script type="text/javascript" src="js/FileSaver.min.js"></script>
	<script type="text/javascript" src="js/parallel.js"></script>
	<script type="text/javascript" src="js/jsfeat.js"></script>
	<script type="text/javascript" src="js/modmap3d.js"></script>

	<script>
	var previousdata = [];
	var labelColors = ['blue', 'red', 'green', 'orange', 'magenta', 'yellow', 'brown', 'lime', 'gray', 'pink','cyan','seagreen', 'olive', 'purple', 'black'];
	var fastaFilesLoaded = [];
	var defaultInputType = "ncbi";
	var offlineCompMinSize = 2000;
	var dev = false;

	function geturlparamvalue(name) {
		name = name.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
		let regexS = "[\\?&]" + name + "=([^&#]*)";
		let regex = new RegExp(regexS);
		let results = regex.exec(window.location.href);
		if(results!=null){
			return results[1];
		}else{
			return -1;
		}
	}

	$(document).ready(function(){
		
		if(geturlparamvalue("dev")=="true"){
			console.log("Dev preview is enabled..");
			dev = true;
		}

		$("#goto1").click(function(){$('#guide-tabs li:eq(1) a').tab('show');});
		$("#goto2").click(function(){$('#guide-tabs li:eq(2) a').tab('show');});
		$("#goto3").click(function(){$('#guide-tabs li:eq(3) a').tab('show');});
		
		$("#kmerslen").val('9');
		$("#getTaxa").val('0');

		$("#buildmapmenucontent").slideUp();
		$("#extendmapmenucontent").slideUp();
		$("#inputdatamenu").slideUp();
		$("#computemenu").slideUp();
		$("#hotkeyslist").slideUp();
		$("#fastafiles").hide();
		
		$("#hotkeysshowhide").click(function(){
			$("#hotkeyslist").slideToggle();
		});

		$("#allInputDataShow").click(function(){
			$("#allInputData").slideToggle();
		});

		$("#buildmapmenuheader").click(function(){
			// console.log("build");
			$("#buildmapmenucontent").slideDown();
			$("#inputdatamenu").slideDown();
			$("#extendmapmenucontent").slideUp();
			$("#computemenu").slideUp();
			$("#numofsets").val(0);
			$("#numofsets").trigger("change");
			// $("#selectnumofsets").css({"border-color": "red", "border-width":"5px", "border-style":"solid"});
		});

		$("#extendmapmenuheader").click(function(){
			// console.log("extend");
			$("#buildmapmenucontent").slideUp();
			$("#inputdatamenu").slideUp();
			$("#extendmapmenucontent").slideDown();
			$("#computemenu").slideUp();
		});


		////////////////////////////
		// [tab] BUILD-EXTEND MAP //
		////////////////////////////

		function load_dataset(dataset){
			$('#allInputData').slideDown();
			$("#file_contents").empty();
			$("#file_contents").hide();
			$("#numofsets").val(dataset.groups.length);
			$("#numofsets").trigger("change");
			for(let i=0; i<dataset.groups.length; i++){
				let cur_group = dataset.groups[i];
				$("#set"+i).html(cur_group.accession_numbers.join(","));
				$("#nameset"+i).val(cur_group.name);
				$("#colorset"+i).val(cur_group.color);		
			}
			$("#stepstatus").hide();	
			$("#mainprogressbar").hide();
			$("#computemenu").slideUp();
		}

		$("#dataset1").click(function(){
			( async () => {
				let utils =  new FileUtils();
				let dataset = await utils.fetch_file("./maps/samplemaps/sample1.json");
				console.log("dataset=", dataset);
				load_dataset(dataset);
			})();
		});  

		$("#dataset2").click(function(){
			( async () => {
				let utils =  new FileUtils();
				let dataset = await utils.fetch_file("./maps/samplemaps/sample2.json");
				console.log("dataset=", dataset);
				load_dataset(dataset);
			})();
		});

		$("#dataset3").click(function(){
			( async () => {
				let utils =  new FileUtils();
				let dataset = await utils.fetch_file("./maps/samplemaps/sample3.json");
				console.log("dataset=", dataset);
				load_dataset(dataset);
			})();
		});
		
		$("#loadrefmap").click(function(){
			( async () => {
				console.log("Loading RefMap: ", $("#extendmapid").val());
				$("#inputdatamenu").slideDown();

				let utils =  new FileUtils();
				let refmap_contents = await utils.fetch_file("./maps/refmaps/refmap_"+$("#extendmapid").val()+".json")
				console.log(refmap_contents);
					
				$('#allInputData').slideDown();
				$("#file_contents").empty();
				$("#file_contents").hide();
				$("#numofsets").val(refmap_contents.groups.length + 1);
				$("#numofsets").trigger("change");

				for(let i=0; i<refmap_contents.groups.length; i++){
					let cur_group = refmap_contents.groups[i];
					$("#set"+i).html(cur_group.accession_numbers.join(","));
					$("#nameset"+i).val(cur_group.name);
					$("#colorset"+i).val(cur_group.color);		
				}

				$("#nameset"+refmap_contents.groups.length).val("NEW_points");
				$("#colorset"+refmap_contents.groups.length).val("black");	
				if( defaultInputType == "ncbi"){
					$("#set"+refmap_contents.groups.length).html($("#extrancbi").val());
				}
				if( defaultInputType == "fasta"){
					let tmpOut = [];
					for(let i=0; i<fastaFilesLoaded.length; i++){
						tmpOut.push("fastaFile"+String(i+1));
					}
					// console.log(tmpOut);
					$("#set"+refmap_contents.groups.length).html(tmpOut.join(","));
				}

				$("#stepstatus").hide();
				$("#mainprogressbar").hide();
				$("#computemenu").slideUp();
			})();
		});

		$('input[type=radio][name=extrapointstype]').change(function() {
			if (this.value == 'ncbi') {
				defaultInputType = "ncbi";
				console.log("ncbi selected..");
				$("#extrancbi").show();
				$("#fastafiles").hide();
				$("#fastafilesinfo").hide();
				$("#inputdatamenu").slideUp();
				$("#computemenu").slideUp(); 
				$("#loadrefmap").show();   	
			}
			else if (this.value == 'fasta') {
				defaultInputType = "fasta";
				console.log("fasta selected..");
				$("#extrancbi").hide();
				$("#fastafiles").show();
				$("#fastafilesinfo").show();
				$("#inputdatamenu").slideUp();
				$("#computemenu").slideUp();
				$("#loadrefmap").hide();

				// Check for the various File API support.
				if(window.File && window.FileReader && window.FileList && window.Blob){
					console.log("everything ok");
				}else{
					console.log('The File APIs are not fully supported in this browser.');
					$('#fastafiles').hide()
					$('#fastafilesinfo').html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><span class="sr-only">Error:</span> ERROR <strong>The File APIs are not fully supported in this browser.</strong> Please use another browser or input NCBI accession numbers instead.</div>');
					$("#inputdatamenu").slideUp();
					$("#computemenu").slideUp(); 
				}

				if(fastaFilesLoaded.length >0){
					$("#loadrefmap").show();
				} 
			}
		});
	
		$("#fastafiles").change(function(evt){
			console.log("Loading fasta files metadata..");
			let files = evt.target.files; 
			fastaFilesLoaded = files;
			console.log("fastaFilesLoaded=", fastaFilesLoaded);
			let output = [];
		    for (let i = 0, f; f = files[i]; i++) {
		      output.push('<li><strong>', escape(f.name), '</strong> (', f.type || 'n/a', ') - ', f.size, ' bytes, last modified: ', f.lastModifiedDate ? f.lastModifiedDate.toLocaleDateString() : 'n/a','</li>');
		    }
		    $("#fastafilesinfo").html('<ul>' + output.join('') + '</ul>');
		    $("#loadrefmap").show();
		});

		$("#kmerslen").change(function(){
			fcgrRes = parseInt($("#kmerslen").val());
			console.log("NEW fcgrRes=",fcgrRes);
		});

		$("#getTaxa").change(function(){
			getTaxa = parseInt($("#getTaxa").val());
			console.log("NEW getTaxa=", getTaxa);
		});

		$("#numofsets").change(function(){
			if($("#numofsets").val() == 0){
				$("#numofsets").css({"border-color": "red", "border-width":"5px", "border-style":"solid"});
			}else{
				// DEFAULT VALUES as reported in chrome by jquery
				// $("#numofsets").css("border-color");
				// $("#numofsets").css("border-width");
				// $("#numofsets").css("border-style");
				$("#numofsets").css({"border-color": "rgb(169, 169, 169)", "border-width":"0.666667px", "border-style":"solid"});
			}
			
			let colorList = '';
			for(let i=0; i<labelColors.length; i++){
				colorList += '<option value="'+labelColors[i]+'">'+ labelColors[i]+'</option>';
			}

			// save previous accessions used
			previousdata = [];
			for(let i=0; i<15; i++){
				if($("#set"+i).html() != undefined){
					previousdata.push([ $("#nameset"+i).val(), $("#colorset"+i).val(), $("#set"+i).html() ]);
				}
			}
			console.log("previous_accessions",previousdata);


			console.log("numofsets=",$("#numofsets").val());
			$("#sets").html('');
			for(let i=0; i<$("#numofsets").val(); i++){
				$("#sets").html($("#sets").html()+'<p>GroupName: <input type="text" size="5" id="nameset'+i+'" value="Set['+i+']"> GroupColor: <select id="colorset'+i+'">'+colorList+'</select><br>Insert comma separated NCBI accession numbers:<br><textarea id="set'+i+'" autofocus  rows="2" cols="40"></textarea></p>');
			}
			for(let i=0; i<$("#numofsets").val(); i++){$("#colorset"+i).val(labelColors[i]);}

			//restore saved accessions
			for(let i=0; i<previousdata.length; i++){
				if($("#set"+i).html() != undefined){
					$("#nameset"+i).val(previousdata[i][0]);
					$("#colorset"+i).val(previousdata[i][1]);
					$("#set"+i).html(previousdata[i][2]);
				}
			}

			if($("#numofsets").val()>0){
				$("#buildmap").show();
			}else{
				$("#buildmap").hide();
			}
		});

		$("#buildmap").click(function(){
			$("#computemenu").slideDown();
			$('#allInputData').slideUp();
			$('#allInputDataMinimized').slideDown();

			$("#file_contents").empty();
			$("#file_contents").hide();
			$("#mainprogressbar").show();
			$("#stepstatus").show();
			$("#stepstatus").html('Step 1 of 3: (NCBI + FCGRs)');
			$('#progress').html('0%'); 
			
			(async () => {

				let cur_dataset = {
					"description": "user-built map",
					"kmer": parseInt($("#kmerslen").val()),
					"enable_taxa_info": ($("#getTaxa").val() == "1"),
					"groups": []
				}
				
				for(let i=0; i<$("#numofsets").val(); i++){
					console.log(i,$("#set"+i).val().split(","));
					let tmpSet = $("#set"+i).val().split(",");
					if(tmpSet[0]!=""){
						cur_dataset["groups"].push({
							"name": $("#nameset" + i).val(),
							"color": $("#colorset" +i ).val(),
							"accession_numbers": [],
						});
						cur_dataset["groups"][i]["accession_numbers"] = tmpSet;
					}
				}
				console.log(JSON.stringify(cur_dataset));
				
				let cur_modmap = new MoDMap3D(cur_dataset);

				if(cur_modmap.accession_numbers.length < 5){
					let alert_msg = new UIUtils().build_alert('<strong>Found ' + cur_modmap.accession_numbers.length + ' sequences.</strong> Please enter at least 5 sequences and try again.', 'danger');
					$('#progress').html(alert_msg);
					return;
				}
				if(cur_modmap.accession_numbers.length > offlineCompMinSize ){
					let alert_msg_raw = "Your input contains more than ["+offlineCompMinSize+"] sequences. This requires that the last step (MDS) has to be computed offline using either Mathematica or Python. Files and code will be provided to you once you reach that point.";
					let alert_msg_html = new UIUtils().build_alert(alert_msg_raw, 'info');
					$('#stepstatus').html( $('#stepstatus').html() + alert_msg_html);
					alert(alert_msg_raw);
				}

				let step1A = new Date().getTime();
				await cur_modmap.fetch_sequences_and_build_fcgrs(fastaFilesLoaded, "progress");
				let step1B = new Date().getTime();
				$("#stepstatus").html('\
					<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</strong></p>\
					<p>Step 2 of 3: (AID)</p>');
				$('#progress').html('0%');


				let step2A = new Date().getTime();
				await cur_modmap.compute_distance_matrix();
				let step2B = new Date().getTime();
				$("#stepstatus").html('\
					<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</p>\
					<p>Step 2 (AID) DONE in ~'+(step2B-step2A)/1000+' sec</p></strong>\
					<p>Step 3 of 3: (MDS)</p>');
				$('#progress').html('Please wait.. (browser may freeze for a while) <img src="img/loading.gif" width="30" height="30"/>');
            

				if(cur_modmap.accession_numbers.length > offlineCompMinSize ){
					$('#progress').html('<p>Please continue computation offline. <br> <strong>Instructions:</strong>\
						<ul><li>Download zip file (It is downloaded automatically)</li>\
						<li>Extract zip contents to a directory of your choice</li>\
						<li>Download either <a href="./extra/mds.nb" target="_blank">MDS-Mathematica</a> or <a href="./extra/mds.py" target="_blank">MDS-Python</a> code and place it in the same directory</li>\
						<li>Run the code, you should get an output file [json_mapfile.json]</li>\
						<li>Open [json_mapfile.json], copy its content and paste it <a href="./extra/input.html" target="_blank">here</a></li>\
						</ul></p>');
					console.log('COPY/PASTE MATHEMATICA');

					cur_modmap.toOfflineComputation();
					return;
				}
				console.log("here");
				console.log(cur_modmap.dist_matrix);

				let step3A = new Date().getTime();
				await cur_modmap.multidimensional_scaling();
				let step3B = new Date().getTime();
				$("#stepstatus").html('\
					<p><strong>Step 1 (NCBI + FCGRs) DONE in ~'+(step1B-step1A)/1000+' sec</p>\
					<p>Step 2 (AID) DONE in ~'+(step2B-step2A)/1000+' sec</p>\
					<p>Step 3 (MDS) DONE in ~'+(step3B-step3A)/1000+' sec</p></strong>');
				$("#stepstatus").html($("#stepstatus").html()+'<p><a href="load.html?mapid=local'+cur_modmap.id+'" target="_blank">Show MoDMap</a>');
				$("#stepstatus").html($("#stepstatus").html()+' or <a href="#viewfile" id="viewfile">view contents</a>');
				$("#stepstatus").html($("#stepstatus").html()+'</p>');
				$("#mainprogressbar").hide();

				let file_utils = new FileUtils();
				let map_key = "local" + cur_modmap.id;
				let map_contents = cur_modmap.toTextFile();
				console.log("map_key=", map_key);

				file_utils.put(map_key, map_contents);


				$("#file_contents").html('<pre>' + file_utils.get(map_key) + '</pre>');
				$("#file_contents").hide();

				$("#viewfile").click(function(){
					$("#file_contents").toggle();
				});
				
			})();
		});


		///////////////////////
		// [tab] COMPUTE AID //
		///////////////////////

		$("#aid1").click(function(){
			$("#seq0").empty();
			$("#seq0").val("NC_012920");
			$("#seq1").empty();
			$("#seq1").val("NC_001643");
			$("#computedist").click();	
		});

		$("#aid2").click(function(){
			$("#seq0").empty();
			$("#seq0").val("NC_012920");
			$("#seq1").empty();
			$("#seq1").val("NC_011137");
			$("#computedist").click();				
		});  
		
		$("#computedist").click(function(){
			(async() =>{
				console.log("AID start");
				$("#aidinfo").empty();
				$("#aidinfomenu").slideDown();
				$("#aidinfo").html('<img src="img/loading.gif" width="30" height="30"/>');
				
				let kmer_length = parseInt($("#aid-kmer-size").val());
				let tmp_dataset = {
					"description": "No_Description",
					"kmer": kmer_length,
					"enable_taxa_info": false,
					"groups": [
						{
							"name": "Strepsirrhini",
							"color": "green",
							"accession_numbers": [$("#seq0").val(), $("#seq1").val()]
						},
					]
				}
				console.log(tmp_dataset);
				
				if((!Number.isInteger(kmer_length)) || (kmer_length < 2) || (kmer_length > 9)){
					$('#aidinfo').html(new UIUtils().build_alert('<strong>ERROR</strong>, kmer size must be an integer  2 <= k <= 9', 'danger'));
					throw Error("kmer size must be an integer  2 <= k <= 9");
				}
				
				let tmp_modmap = new MoDMap3D(tmp_dataset);
				try{
					await tmp_modmap.fetch_sequences_and_build_fcgrs([]);
				}catch (err){
					$('#aidinfo').html(new UIUtils().build_alert('<strong>ERROR</strong>, ' + err.message, 'danger'));
					throw err;
				}

				let aid_distance = tmp_modmap._compute_AID_distance(
					tmp_modmap.all_sequences_info[0]['fcgr'], 
					tmp_modmap.all_sequences_info[1]['fcgr'], 9);
				console.log(aid_distance);

				let seq_1_info = tmp_modmap.all_sequences_info[0];
				let seq_2_info = tmp_modmap.all_sequences_info[1];
				let aid_info_text = "<strong>NCBI: ["+seq_1_info['accession_number']+"]</strong> - Length: "+seq_1_info['sequence_length']+"<br>Header: "+seq_1_info['header']+"<br><br>" + 
								"<strong>NCBI: ["+seq_2_info['accession_number']+"]</strong> - Length: "+seq_2_info['sequence_length']+"<br>Header: "+seq_2_info['header']+"<br><br>" +
								"<strong>AID (k=" + kmer_length + "): </strong>"+ String(aid_distance.toFixed(5));
				$("#aidinfo").html(aid_info_text);

				console.log("AID end");
			})();

		});
		

		///////////////////////
		// [tab] EXPLORE MAP //
		///////////////////////
		
		// POPULATE LEFT MENU WITH CATEGORIES AND MAPS IN THEM
		( async () => {
			let utils =  new FileUtils();
			let maps_collection = await utils.fetch_file("./extra/maps.json");
			console.log("maps_collection=", maps_collection);

			let maps_collection_groups = [];
			for(let i=0; i<maps_collection.maps.length; i++){
				maps_collection_groups.push(maps_collection.maps[i]);
			}
			if(dev){
				for(let i=0; i<maps_collection.dev.length; i++){
					maps_collection_groups.push(maps_collection.dev[i]);
				}
			}
			// console.log(maps_collection_groups);

			let outcome='';
			for(let i=0; i<maps_collection_groups.length; i++){
				let maps_group = maps_collection_groups[i];
				
				outcome += '<div class="panel-heading">\
							<h4 class="panel-title">\
							<a data-toggle="collapse" data-parent="#modmap-accordion-' + i + '" href="#modmap-collapse-' + i + '">' +  maps_group.category + '&nbsp;&nbsp;<span class="badge">' + maps_group.files.length+'</span></a>\
							</h4></div>';
				let tmp = '<div id="modmap-collapse-' + i + '" class="panel-collapse collapse out"><div class="panel-body"><ul class="list-group large-group">';
				for(let j=0; j<maps_group.files.length; j++){
					tmp += '<li class="list-group-item" id="' + maps_group.files[j] + '"><a href="#">' + maps_group.files[j] + '</a></li>';
				}
				tmp+='</ul></div></div>';
				outcome+=tmp;
			}

			$('#listgroupsofmaps').html(outcome);
		})();
		
		// LOAD PREVIEW CONTENT FOR EACH FILENAME
		$("body").on("click", ".list-group-item", function (event) {
            event.preventDefault();
			
			( async () => {
				$("#filecontent").html('<img src="img/loading.gif" width="50" height="50"/>');
				console.log('Loading = '+$(this).text());
				let map_id = $(this).text();

				let utils =  new FileUtils();
				let map_contents = await utils.fetch_file("./maps/" + map_id, true);
				// console.log(map_contents);

				let resfetch=map_contents.split("\n");
				if(resfetch[6].indexOf("#") > -1){
					// NEW TYPE OF LEGEND WITH INFO SPLIIT BY #
					console.log("Found legend, split by #");
					let sublegends = resfetch[6].split("#");
					let subleg = "<div>&bull; "+sublegends[0]+"</div>";
					subleg += "<div>&bull; "+sublegends[1]+"</div>";
					subleg += "<div>&bull; "+sublegends[2]+"</div>";
					subleg += "<div>&bull; "+sublegends[3]+"</div>";
					subleg += "<div>&bull; Avg.Seq.Length: "+sublegends[4]+"</div>";
					$("#filecontent").html('<p id="mapId"><strong>'+resfetch[5]+'</strong></p><p>'+subleg+'</p><a href="load.html?mapid='+map_id+'" target="_blank">Show Map</a><p></p>');
				}else{
					// BACKWARD COMPATIBILITY LEGEND
					console.log("BACKWARD COMPATIBILITY LEGEND");
					$("#filecontent").html('<p id="mapId"><strong>'+resfetch[5]+'</strong></p><p>'+resfetch[6]+'</p><a href="load.html?mapid='+map_id+'" target="_blank">Show Map</a><p></p>');
				}
			})();
		});


		// LEGACY, NOT SURE WHAT THE USE CASE IS. Leaving it for now.
		let tmpDataset = geturlparamvalue('dataset');
		if(tmpDataset!=-1 && tmpDataset.slice(0,5)=='local'){
			newDataset = localStorage.getItem("dataset"+tmpDataset.slice(5));
			$("#buildmapmenuheader").click();
			$('#guide-tabs li:eq(2) a').tab('show');
			$("#file_contents").empty();
			$("#file_contents").hide();
			$("#numofsets").val('1');
			$("#numofsets").trigger("change");
			$("#set0").html(newDataset);
			$("#stepstatus").hide();
			$("#mainprogressbar").hide();
		}
	
	});

	</script>

	<style>
	body {
		margin-bottom: 60px;
	}
	#listgroupsofmaps{
		margin: 10px;
	}
	#filecontent{
		margin: 10px;
	}
	</style>
</head>
<body>

	<div class = "page-header">
		<h1 class="text-center">MoDMaps3D</h1>
	</div>

	<div class="container">
		<ul class="nav nav-tabs" id="guide-tabs">
			<li class="active"><a data-toggle="tab" href="#home">Home</a></li>
			<li><a data-toggle="tab" href="#menu1">Explore a Map</a></li>
			<li><a data-toggle="tab" href="#menu2">Build or Extend a Map</a></li>
			<li><a data-toggle="tab" href="#menu3">Compute a distance</a></li>
			<li><a data-toggle="tab" href="#menu4">About</a></li>
		</ul>

		<div class="tab-content">
			<div id="home" class="tab-pane fade in active">
				<h3>Welcome to MoDMaps3D!</h3>
				<p>Here are some of the things you can do with three dimensional Molecular Distance Maps.</p>

				<dl class="dl-horizontal">
					<dt><a href="#" id="goto1">Explore a Map</a></dt>
					<dd>Explore available built-in MoDMaps. These maps provide extensive information for each genomic sequence involved in each map, along with complementary information.</dd>
					<dt><a href="#" id="goto2">Build or Extend a Map</a></dt>
					<dd>Build a new MoDMap from scratch tailored to your dataset. Include up to 12 categories, with as many sequences you need in each of them. Alternatively, extend a MoDMap adding your points. You can select a "base" map from a variety of available maps and then add your DNA sequences in it.</dd>
					<dt><a href="#" id="goto3">Compute a distance</a></dt>
					<dd>Compute the distance between any two DNA sequences, given as NCBI accession numbers.</dd>
				</dl>
				<div id="fastafile"></div>
			</div>
			<div id="menu1" class="tab-pane fade">
				<h3>Explore available MoDMaps</h3>
				<p>Here you can find a list of available built-in MoDMaps to explore. Select a map from sidebar, preview its content, and click on "Show Map" to see it in action.</p>
				
				<div class="col-xs-5">
					<div id="listgroupsofmaps" class="panel panel-default">
					</div>
				</div>
				<div class="col-xs-6 " id="filecontent">
				</div>  
			</div>
			<div id="menu2" class="tab-pane fade">
				<!-- style="padding-right:0px; border-right: 1px solid #ccc;" -->
				<div class="col-xs-10" id="buildmapview" >
					<h3>Build your map, for your dataset</h3>
					
					<!-- <div class="panel panel-default">
					  <div class="panel-heading">Panel with panel-default class</div>
					  <div class="panel-body">Panel Content</div>
					</div> -->

					<div class="panel panel-default">
						<div class="panel-heading"><label>Choose an option to continue</label></div>
						<div class="panel-body">
						<label><input type="radio" id="buildmapmenuheader" name="mapmenu"/> Build a MoDMap from your dataset</label><br>
						<label><input type="radio" id="extendmapmenuheader" name="mapmenu"/> Extend an existing MoDMap</label>
						</div>
					</div>

					<div class="panel panel-default" id="buildmapmenucontent">
						<div class="panel-heading"><label>Info</label></div>
						<div class="panel-body">
							<p>If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>. Samples you can try: 
							<a href="#" id="dataset1">[1] 62 mtDNA primates</a> - 
							<a href="#" id="dataset2">[2] 112 mtDNA amphibia</a> - 
							<a href="#" id="dataset3">[3] 39 HIV-1 genomes</a> 
							</p>
						</div>
					</div>

					<div class="panel panel-default" id="extendmapmenucontent">
						<div class="panel-heading"><label>Info</label></div>
						<div class="panel-body">
							<p>Extend a map by adding new points in a set of predefined MoDMaps. The new points can be given as comma separated accession numbers from NCBI. If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>.</p>


							<p>
							<dl class="dl-horizontal">
							<dt>Background map:</dt>
							<dd><select id="extendmapid">
								<option value="Animalia_mtDNA_ClassAmphibia">Animalia_mtDNA_ClassAmphibia</option>
								<!-- INSECTS REMOVED BECAUSE OF TOO MANY SUBSETS 21>>10 -->
								<!-- <option value="Animalia_mtDNA_ClassInsecta">Animalia_mtDNA_ClassInsecta</option> -->
								<option value="Animalia_mtDNA_ClassMammalia">Animalia_mtDNA_ClassMammalia</option>
								<option value="Animalia_mtDNA_OrderPrimates">Animalia_mtDNA_OrderPrimates</option>
								<option value="Animalia_mtDNA_Vertebrata">Animalia_mtDNA_Vertebrata</option>
								<option value="Animalia_mtDNA_Ins_Mam_Amph">Animalia_mtDNA_Ins_Mam_Amph</option>
								<option value="Fungi_mtDNA">Fungi_mtDNA</option>
								<option value="Plants_mtDNA">Plants_mtDNA</option>
								<option value="Protists_mtDNA">Protists_mtDNA</option>
								<option value="Protists_ptDNA">Protists_ptDNA</option>
								<option value="Archaea_nDNA">Archaea_nDNA</option>
								<option value="Bacteria_nDNA">Bacteria_nDNA</option>
								<option value="Viruses_HIV1">Viruses_HIV1</option>
								<option value="Haplogroups_All">Haplogroups_All</option>
								<option value="chimeras1">chimeras1</option>
								<option value="chimeras2">chimeras2</option>
							</select><br></dd>
							<dt>Extra Sequences:</dt>
							<dd>
								<label><input type="radio" name="extrapointstype" value="ncbi" checked="checked">
								Insert comma separated NCBI accession numbers</label><br>
								<label><input type="radio" name="extrapointstype" value="fasta">
								Load FASTA files</label><br>
								<textarea id="extrancbi" rows="2" cols="40"></textarea>
								<input type="file" id="fastafiles" name="fastafiles[]" multiple >
								<output id="fastafilesinfo"></output>

							</dd>

							<dt>
							<!-- Enter comma separated NCBI accessions --OR-- load FASTA files<br>
							</dd> -->
							</dl></p>

							<p><input type="submit" id="loadrefmap" value="Continue"/></p>
						</div>
					</div>

					<div class="panel panel-default" id="inputdatamenu">
						<div class="panel-heading"><label>Settings of MoDMap selected</label></div>
						<div class="panel-body">
							<p><div id="allInputDataMinimized">
								<a href="#" id="allInputDataShow">Show/Hide input data</a>
							</div></p>

							<div id="allInputData">
								<dl class="dl-horizontal">
									<dt>Length of k-mers:</dt>
									<dd>
										<select id="kmerslen">
											<option value="3">3 bp</option>
											<option value="4">4 bp</option>
											<option value="5">5 bp</option>
											<option value="6">6 bp</option>
											<option value="7">7 bp</option>
											<option value="8">8 bp</option>
											<option value="9">9 bp</option>
											<option value="10">10 bp</option>
											<option value="11">11 bp</option>
											<option value="12">12 bp</option>
										</select>
									</dd>
									<dt>Get Taxonomy:</dt>
									<dd>
										<select id="getTaxa">
											<option value="0">No (Fast)</option>
											<option value="1">Yes (Slow)</option>
										</select>
									</dd>
									<dt>Number of groups:</dt>
									<dd id="selectnumofsets">
										<select id="numofsets">
											<option value="0">Choose</option>
											<option value="1">1</option>
											<option value="2">2</option>
											<option value="3">3</option>
											<option value="4">4</option>
											<option value="5">5</option>
											<option value="6">6</option>
											<option value="7">7</option>
											<option value="8">8</option>
											<option value="9">9</option>
											<option value="10">10</option>
											<option value="11">11</option> 
											<option value="12">12</option>
											<option value="13">13</option>
											<option value="14">14</option>
											<option value="15">15</option> 
										</select>
									</dd>
								</dl>

								<div id="sets"></div>
							</div>

							<p><input type="submit" id="buildmap" value="Compute MoDMap" style="display:none;" /></p>
						</div>
					</div>

					<div class="panel panel-default" id="computemenu">
					<div class="panel-heading"><label>Computation Results</label></div>
					<div class="panel-body">
						<p id="stepstatus"></p>
						<div id="mainprogressbar" style="display:none;">
						  <div id="progress"></div>
						</div>
						<p id="file_contents"></p>
					</div>
					</div>

				</div>
			</div>
			<div id="menu3" class="tab-pane fade">
				<h3>Compute a distance</h3>
				<p>Compute approximated information distance (AID) between any pair of sequences, giving as input their NCBI accession numbers. AID distance has a value in [0,1]. If you do not know the accession number of a species, you can search directly in NCBI from <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Homo+sapiens" target="_blank">here</a>.</p>

				<div class="panel panel-default">
					<div class="panel-heading"><label>Settings</label></div>
					<div class="panel-body">
						<p>Samples you can try:
						<a href="#" id="aid1">[1] H.sapiens vs P.troglodytes</a> - 
						<a href="#" id="aid2">[2] H.sapiens vs H.neanderthalensis</a></p>
						Compute AID distance between (input NCBI accession numbers): 
						<ol>
							<li><input id="seq0" type="text" size="10" autofocus ></li>
							<li><input id="seq1" type="text" size="10" ></li>
						</ol>	
						For kmer size:
						<input id="aid-kmer-size" type="number" min="2" max="12" step="1" value="9"><br><br>
						<input type="submit" id="computedist" value="Compute AID distance" />
					</div>
				</div>

				<div class="panel panel-default"  id="aidinfomenu" style="display:none;">
					<div class="panel-heading"><label>Computation Results</label></div>
					<div class="panel-body"  id=aidinfo></div>
				</div>
			</div>
			<div id="menu4" class="tab-pane fade">
				<p>
				<dl class="dl-horizontal">
					<dt>Coded by:</dt>
					<dd>Rallis Karamichalis, 2016</dd>
					<dt>Github Code:</dt>
					<dd><a href="http://github.com/rallis/MoDMaps3D/" target="_blank">http://github.com/rallis/MoDMaps3D/</a></dd>
					<dt>HotKeys:</dt>
					
					<dd>
						<div id="hotkeysshowhide"><a href="#">Show/Hide full list</a></div>
						<div id="hotkeyslist">
						While on canvas:
						<ul>
						<li>A = Move camera Left</li>
						<li>D = Move camera Right</li>
						<li>W = Move camera Up</li>
						<li>S = Move camera Down</li>
						<li>E = Zoom camera In</li>
						<li>Q = Zoom camera Out</li>
						<li>Ctrl + LeftArrow = Rotate Left</li>
						<li>Ctrl + UpArrow = Rotate Up</li>
						<li>Ctrl + RightArrow = Rotate Right</li>
						<li>Ctrl + DownArrow = Rotate Down</li>
						<li>Left Mouse Click + Drag = Rotate to any direction</li>
						</ul>

						General Shortcuts:
						<ul>
						<li>Ctrl + F = Search field</li>
						<li>Ctrl + Alt + R = Reset Camera view, keeping rotation</li>
						<li>Ctrl + Alt + M = Toggle visibility of right menu</li>
						<li>Ctrl + Alt + N = Toggle visibility of left menu</li>
						</ul>
						</div>
					</dd>
					
				</dl>
				</p>	
			</div>
		</div>	
	</div>

</body>
</html>
